test/apps/iodemo/io_demo.cc

```diff
@@ -67,40 +68,72 @@ typedef struct {
 template<class T>
 class MemoryPool {
 public:
-    MemoryPool(size_t buffer_size) :
+    MemoryPool(size_t buffer_size, size_t offcache) :
         _num_allocated(0), _buffer_size(buffer_size) {
+
+        for (size_t i = 0; i < offcache; ++i) {
+            _offcache_queue.push(get_free());
+        }
     }
 
     ~MemoryPool() {
-        if (_num_allocated != _freelist.size()) {
+        while (!_offcache_queue.empty()) {
+            put_free(_offcache_queue.front());
+            _offcache_queue.pop();
+        }
+
+        if (_num_allocated != _free_stack.size()) {
             LOG << "Some items were not freed. Total:" << _num_allocated
-                << ", current:" << _freelist.size() << ".";
+                << ", current:" << _free_stack.size() << ".";
+        }
+
+        for (size_t i = 0; i < _free_stack.size(); i++) {
+            delete _free_stack[i];
+        }
+    }
+
+    inline T* get() {
+        if (_offcache_queue.empty()) {
+            return get_free();
+        } else {
+            T* item = _offcache_queue.front();
+            _offcache_queue.pop();
+            _offcache_queue.push(get_free());
+            return item;
         }
+    }
 
-        for (size_t i = 0; i < _freelist.size(); i++) {
-            delete _freelist[i];
+    inline void put(T* item) {
+        if (_offcache_queue.empty()) {
+            put_free(item);
+        } else {
+            put_free(_offcache_queue.front());
+            _offcache_queue.pop();
+            _offcache_queue.push(item);
         }
     }
 
-    T * get() {
-        T * item;
-        
-        if (_freelist.empty()) {
+private:
+    inline T* get_free() {
+        T* item;
+
+        if (_free_stack.empty()) {
             item = new T(_buffer_size, *this);
             _num_allocated++;
         } else {
-            item = _freelist.back();
-            _freelist.pop_back();
+            item = _free_stack.back();
+            _free_stack.pop_back();
         }
         return item;
     }
 
-    void put(T * item) {
-        _freelist.push_back(item);
+    inline void put_free(T* item) {
+        _free_stack.push_back(item);
     }
 
 private:
-    std::vector<T*> _freelist;
+    std::vector<T*> _free_stack;
+    std::queue<T*>  _offcache_queue;
```

Comment:
is std::queue doing malloc/free? maybe std::deque?

Reply:
default underlying container of the queue is deque, see https://en.cppreference.com/w/cpp/container/queue

Reply:
ok